src/Adapter/OrmAdapter.php

```diff
@@ -0,0 +1,312 @@
+use Doctrine\Common\Util\ClassUtils;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Symfony\Cmf\Bundle\CoreBundle\Translatable\TranslatableInterface;
+use Symfony\Cmf\Bundle\RoutingAutoBundle\Doctrine\Orm\AutoRoute;
+use Symfony\Cmf\Component\RoutingAuto\AdapterInterface;
+use Symfony\Cmf\Component\RoutingAuto\Model\AutoRouteInterface;
+use Symfony\Cmf\Component\RoutingAuto\UriContext;
+
+/**
+ * Adapter for ORM.
+ *
+ * @author WAM Team <develop@wearemarketing.com>
+ */
+class OrmAdapter implements AdapterInterface
+{
+    const TAG_NO_MULTILANG = 'no-multilang';
+
+    const ID_PLACEHOLDER = '%ID_PLACEHOLDER%';
+
+    const REQUEST_LOCALE_ATTRIBUTE = '_locale';
+
+    /**
+     * @var EntityManagerInterface
+     */
+    private $em;
+
+    /**
+     * @var string
+     */
+    private $autoRouteFqcn;
+
+    /**
+     * @var ObjectRepository
+     */
+    private $repository;
+
+    /**
+     * @param EntityManagerInterface $em
+     * @param string                 $autoRouteFqcn The FQCN of the AutoRoute document to use
+     */
+    public function __construct(EntityManagerInterface $em, $autoRouteFqcn)
+    {
+        $this->em = $em;
+
+        $reflection = new \ReflectionClass($autoRouteFqcn);
+        if (!$reflection->isSubclassOf('Symfony\Cmf\Component\RoutingAuto\Model\AutoRouteInterface')) {
+            throw new \InvalidArgumentException(sprintf('AutoRoute documents have to implement the AutoRouteInterface, "%s" does not.', $autoRouteFqcn));
+        }
+
+        $this->autoRouteFqcn = $autoRouteFqcn;
+
+        $this->repository = $this->em->getRepository($this->autoRouteFqcn);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getLocales($contentDocument)
+    {
+        // TODO: look for better approach. This is because we are using knp doctrine behaviour lib
+        if ($contentDocument instanceof TranslatableInterface) {
+            return array_keys($contentDocument->getTranslations()->toArray());
+        }
+
+        return [];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function translateObject($contentDocument, $locale)
+    {
+        $contentDocument->setCurrentLocale($locale);
+
+        return $contentDocument->translate(null, false);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generateAutoRouteTag(UriContext $uriContext)
+    {
+        return $uriContext->getLocale() ?: self::TAG_NO_MULTILANG;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function migrateAutoRouteChildren(AutoRouteInterface $srcAutoRoute = null, AutoRouteInterface $destAutoRoute = null)
+    {
+        // It is not implemented tree relationship between routes
+        // TODO: if we implemented tree we can use this method for propageting update.
+        // Indeed, we can try to use this method to update route when depends on other one like /category/product and we change category we have to apdate all products.
```

Comment:
The URL is the tree, so updating the auto route children would be something like selecting all `LIKE "/parent/url%` and updating them

Reply:
good point... I will try to think about and try to do something

Reply:
I am not sure we can do it like you say. For instance, it is not compulsory that the path has to be /{category}/{product}, it could it be /{product}/{category}. Then, if we change category, how could we find paths of product? I am sorry, but I don't find a better example...

Reply:
I think I've found it in the project of an old client. The following path /events/{category-title}/{region-title} is a filter for events. Category and region are auto routable entities and both have two schemas: one for itself (/category/{category-title} and the other one when we mix them (/events/{category-title}/{region-title}).
If we update category, we can do it. But, if we modify region, how could we do it?

Reply:
The URLs are expanded? so you have `/my-category/my-region`. Renaming if you rename `my-category` to `their-category` then you need to migrate all of the "child" regions, so, `/my-category/my-region` becomes `/their-category/my-region` (and same for any other children of category, or?